File: dashboard/lib/entitlements.ts
// Single source of tier truth for the dashboard runtime. MIRRORED, field-for-field,
// by reviewer/entitlements.py (the reviewer/worker consume the same numbers) and
// guarded by tests/test_entitlements_parity.py, which regex-extracts the model ids,
// caps, and monthly allowances from THIS file and asserts equality with the Python
// module. Keep the two in lockstep — the parity test fails loudly on drift.
//
// Pricing table: spec 2026-07-03 "Pricing & tiers". A tier is a monthly compute
// budget; the per-model daily review cap holds cost constant across model choice
// (cheap model → high cap, premium model → low cap). Premium-model access is the
// real differentiator, not review volume.

export const CHEAP_MODEL = "deepseek/deepseek-v4-flash";
export const PREMIUM_MODEL = "anthropic/claude-haiku-4.5";

export type Plan = "standard" | "pro";
export type ModelSlot = "cheap" | "premium";

export interface Entitlement {
  // slot → per-user, per-day cap on jobs entering review on that model.
  stage2Models: Partial<Record<ModelSlot, number>>;
  monthlyResume: number;
  monthlyCover: number;
}

export type EntitlementMap = Record<Plan, Entitlement>;

export const ENTITLEMENTS: EntitlementMap = {
  standard: { stage2Models: { cheap: 400 }, monthlyResume: 30, monthlyCover: 30 },
  pro: { stage2Models: { cheap: 1000, premium: 100 }, monthlyResume: 100, monthlyCover: 100 },
};

// Display-only monthly price (USD), spec "Pricing & tiers". Dashboard billing UI reads
// this so the price is never hardcoded in JSX; the reviewer doesn't need it, so it is
// intentionally TS-only (not mirrored in entitlements.py / the parity test).
//
// NOTE (T1): these are the COMPILE-TIME DEFAULTS. Money-gating call sites read the
// DB-overlaid values via lib/tierConfig.ts loadTierConfig() instead of these constants
// directly, so caps/allowances/prices are tunable without a redeploy. The pure functions
// below accept an optional `ent` map (defaulting to ENTITLEMENTS) so they operate on
// either the compiled defaults or the loaded overlay with identical semantics.
export const PLAN_PRICE_USD: Record<Plan, number> = { standard: 5, pro: 20 };
export const PLAN_LABEL: Record<Plan, string> = { standard: "Standard", pro: "Pro" };

// ── Invite comp plan (user-sent invites, spec 2026-07-13) ────────────────────
// What an invited-but-not-paying user is comped. DB-overridable via
// app_settings.invite_comp_plan (lib/appSettings.ts); this compiled constant is the
// fallback AND the parity-guarded default — tests/test_entitlements_parity.py
// regex-extracts it and asserts equality with reviewer/entitlements.py, so keep the
// bare `export const NAME = "value";` shape.
export type InviteCompPlan = Plan | "none";
export const DEFAULT_INVITE_COMP_PLAN = "standard";

// ── Reasoning effort (résumé / cover-letter generation) ─────────────────────
// TS-ONLY, intentionally NOT mirrored in reviewer/entitlements.py: generation is
// dashboard-only, the reviewer never reads these (same precedent as
// PLAN_PRICE_USD). Kept OUTSIDE the ENTITLEMENTS literal so the parity test's
// regexes never see it. Compile-time constants, not part of the tierConfig DB
// overlay — effort tiers are product shape, not a tunable money cap.
export type ReasoningEffort = "off" | "low" | "medium" | "high";

const EFFORT_ORDER: ReasoningEffort[] = ["off", "low", "medium", "high"];

export const REASONING_EFFORTS: Record<Plan, ReasoningEffort[]> = {
  standard: ["off", "low"],
  pro: ["off", "low", "medium", "high"],
};

/**
 * Call-time clamp (mirrors resolveStage2Model's hard fallback): the requested
 * effort if the plan grants it, otherwise the highest granted level below it —
 * so a Pro→Standard downgrade with a saved "high" degrades to "low", not "off".
 * null plan → "off" (the routes' 402 gate fires before this ever matters).
 */
export function resolveReasoningEffort(
  plan: Plan | null,
  requested: ReasoningEffort,
): ReasoningEffort {
  if (!plan) return "off";
  const allowed = REASONING_EFFORTS[plan];
  for (let i = EFFORT_ORDER.indexOf(requested); i > 0; i--) {
    if (allowed.includes(EFFORT_ORDER[i])) return EFFORT_ORDER[i];
  }
  return "off";
}

export type ReasoningEffortValidation =
  | { ok: true; value: "low" | "medium" | "high" | null }
  | { ok: false; reason: string };

/**
 * Save-time gate for the profile form (mirrors the stage-2 model gate): ""/"off"
 * normalize to null (Off, the stored default), low passes on any plan, medium/high
 * require Pro, anything else (hand-crafted post) is rejected.
 */
export function validateReasoningEffort(
  raw: string,
  plan: Plan | null,
): ReasoningEffortValidation {
  const v = raw.trim().toLowerCase();
  if (!v || v === "off") return { ok: true, value: null };
  if (v !== "low" && v !== "medium" && v !== "high") {
    return { ok: false, reason: `unknown reasoning effort: ${raw}` };
  }
  if (v === "low") return { ok: true, value: "low" };
  if (plan !== "pro") {
    return { ok: false, reason: "Medium and High reasoning effort require the Pro plan." };
  }
  return { ok: true, value: v };
}

// Subscription is valid for 3 extra days past current_period_end so a webhook lag or
// renewal-retry doesn't briefly strand a paying user.
const GRACE_MS = 3 * 24 * 60 * 60 * 1000;

// TRIALING and PREMIUM (defense-in-depth): we do NOT configure Stripe trials today
// (no trial_period_days on any price), so a `trialing` subscription is not expected
// from our own checkout. It means an in-good-standing but UNPAID subscription; we let
// it entitle at Standard so a legitimate future trial keeps working, but clamp it
// BELOW Pro so a zero-cost trial can never unlock Pro's premium-model daily budget —
// the real cost lever. Flip this to true ONLY when a paid-trial product deliberately
// grants full-plan access during the trial window. Mirrored in entitlements.py.
const TRIAL_GRANTS_FULL_PLAN = false;

/** Which entitlement slot a concrete OpenRouter model id maps to (null = neither). */
export function modelSlot(model: string | null | undefined): ModelSlot | null {
  if (model === PREMIUM_MODEL) return "premium";
  if (model === CHEAP_MODEL) return "cheap";
  return null;
}

export interface SubscriptionLike {
  plan: Plan | string | null;
  status: string | null;
  current_period_end: Date | string | null;
}

/**
 * The user's effective plan under the chargeable-beta policy:
 *   - a paying subscriber (status active|trialing AND current_period_end + grace >
 *     now) gets their subscribed plan;
 *   - otherwise a Phase-0 invitee (invited === true) is comped at the configured
 *     comp plan (default Standard) — a deliberate choice so trusted testers keep
 *     working through the beta;
 *   - a stranger with neither gets null (no entitlement → gated everywhere).
 */
export function resolvePlan(
  sub: SubscriptionLike | null,
  invited: boolean,
  now: Date = new Date(),
  compPlan: InviteCompPlan = DEFAULT_INVITE_COMP_PLAN as InviteCompPlan,
): Plan | null {
  if (
    sub &&
    (sub.status === "active" || sub.status === "trialing") &&
    (sub.plan === "standard" || sub.plan === "pro") &&
    sub.current_period_end != null
  ) {
    const cpe =
      sub.current_period_end instanceof Date
        ? sub.current_period_end
        : new Date(sub.current_period_end);
    if (!Number.isNaN(cpe.getTime()) && cpe.getTime() + GRACE_MS > now.getTime()) {
      // Clamp a trialing subscription below Pro (see TRIAL_GRANTS_FULL_PLAN): an unpaid
      // trial entitles at most to Standard so it can't unlock the premium-model budget.
      if (sub.status === "trialing" && !TRIAL_GRANTS_FULL_PLAN && sub.plan === "pro") {
        return "standard";
      }
      return sub.plan;
    }
  }
  // Comped plan is operator-configurable (app_settings.invite_comp_plan); "none"
  // switches comping off entirely — invited users then need a real subscription.
  if (invited && compPlan !== "none") return compPlan;
  return null;
}

/**
 * The stage-2 model the user is entitled to run: the requested model if the plan
 * grants its slot, otherwise the always-available cheap model. null plan → cheap
 * (the reviewer skips null-plan users before this ever matters).
 */
export function resolveStage2Model(
  plan: Plan | null,
  requestedModel: string | null | undefined,
  ent: EntitlementMap = ENTITLEMENTS,
): string {
  if (plan) {
    const slot = modelSlot(requestedModel);
    if (slot && ent[plan].stage2Models[slot] != null) return requestedModel!;
  }
  return CHEAP_MODEL;
}

/** Per-user, per-day review cap for (plan, resolved stage-2 model). null plan → 0. */
export function dailyReviewCap(
  plan: Plan | null,
  model: string,
  ent: EntitlementMap = ENTITLEMENTS,
): number {
  if (!plan) return 0;
  const e = ent[plan];
  const slot = modelSlot(model) ?? "cheap";
  return e.stage2Models[slot] ?? e.stage2Models.cheap ?? 0;
}

/** Monthly generation allowance for a kind. null plan → 0. */
export function monthlyAllowance(
  plan: Plan | null,
  kind: "resume" | "cover",
  ent: EntitlementMap = ENTITLEMENTS,
): number {
  if (!plan) return 0;
  const e = ent[plan];
  return kind === "resume" ? e.monthlyResume : e.monthlyCover;
}

File: reviewer/entitlements.py
"""Single source of tier truth for the Python runtime (reviewer + worker).

MIRRORS dashboard/lib/entitlements.ts field-for-field; tests/test_entitlements_parity.py
regex-extracts the TS values and asserts equality with the constants here. Keep the
two in lockstep. Stdlib only (datetime) so the reviewer can import it anywhere.

Pricing: spec 2026-07-03 "Pricing & tiers".
"""
from datetime import datetime, timedelta, timezone

CHEAP_MODEL = "deepseek/deepseek-v4-flash"
PREMIUM_MODEL = "anthropic/claude-haiku-4.5"

# plan -> {stage2_models: {slot: per-day review cap}, monthly_resume, monthly_cover}
ENTITLEMENTS = {
    "standard": {"stage2_models": {"cheap": 400}, "monthly_resume": 30, "monthly_cover": 30},
    "pro": {"stage2_models": {"cheap": 1000, "premium": 100}, "monthly_resume": 100, "monthly_cover": 100},
}

# 3-day grace past current_period_end (webhook lag / renewal retry). Mirrors GRACE_MS.
_GRACE = timedelta(days=3)

# Mirrors TRIAL_GRANTS_FULL_PLAN in entitlements.ts. Trials are not configured today
# (no Stripe trial_period_days); a `trialing` subscription entitles at most to Standard
# so a zero-cost trial can never unlock Pro's premium-model daily budget. Flip to True
# only when a paid-trial product deliberately grants full-plan access during the trial.
_TRIAL_GRANTS_FULL_PLAN = False


def _pos_int(v):
    """A positive int (caps/allowances) or None. Rejects bool, float, str, <=0.

    bool is an int subclass in Python, so exclude it explicitly.
    """
    if isinstance(v, bool) or not isinstance(v, int) or v <= 0:
        return None
    return v


def overlay_entitlements(rows):
    """Overlay DB tier_settings onto the compiled ENTITLEMENTS, field-by-field (T1).

    `rows` is an iterable of mappings {plan, config} where config is the jsonb value
    (a dict, or anything for a malformed row). Mirrors dashboard/lib/tierConfig.ts:
    every bad/absent field keeps the compiled default; never raises. A DB row may only
    override slots the compiled default already grants (it cannot invent a premium slot
    for Standard). Returns a fresh entitlements map (the compiled one is never mutated).
    """
    out = {plan: {"stage2_models": dict(ENTITLEMENTS[plan]["stage2_models"]),
                  "monthly_resume": ENTITLEMENTS[plan]["monthly_resume"],
                  "monthly_cover": ENTITLEMENTS[plan]["monthly_cover"]}
           for plan in ENTITLEMENTS}
    by_plan = {}
    for r in rows or []:
        try:
            by_plan[r["plan"]] = r["config"]
        except (KeyError, TypeError):
            continue
    for plan in out:
        cfg = by_plan.get(plan)
        if not isinstance(cfg, dict):
            continue
        s2 = cfg.get("stage2Models")
        if isinstance(s2, dict):
            for slot in list(out[plan]["stage2_models"].keys()):
                if slot in s2:
                    cap = _pos_int(s2[slot])
                    if cap is not None:
                        out[plan]["stage2_models"][slot] = cap
        if "monthlyResume" in cfg:
            n = _pos_int(cfg["monthlyResume"])
            if n is not None:
                out[plan]["monthly_resume"] = n
        if "monthlyCover" in cfg:
            n = _pos_int(cfg["monthlyCover"])
            if n is not None:
                out[plan]["monthly_cover"] = n
    return out


def model_slot(model):
    """Entitlement slot for a concrete OpenRouter model id (None = neither)."""
    if model == PREMIUM_MODEL:
        return "premium"
    if model == CHEAP_MODEL:
        return "cheap"
    return None


def resolve_plan(sub, invited, now=None):
    """The user's effective plan under the chargeable-beta policy.

    sub: mapping with keys plan, status, current_period_end (a tz-aware datetime or
    None), or None when the user has no subscription row. invited: server-side
    invite proof. Returns 'standard' | 'pro' | None with semantics identical to
    resolvePlan in entitlements.ts:
      - active|trialing subscription within (current_period_end + 3-day grace) -> its plan
      - else invited (comped Phase-0 beta) -> 'standard'
      - else -> None
    """
    if now is None:
        now = datetime.now(timezone.utc)
    if sub is not None:
        status = sub.get("status")
        plan = sub.get("plan")
        cpe = sub.get("current_period_end")
        if status in ("active", "trialing") and plan in ("standard", "pro") and cpe is not None:
            if cpe + _GRACE > now:
                # Clamp a trialing subscription below Pro (see _TRIAL_GRANTS_FULL_PLAN):
                # an unpaid trial entitles at most to Standard, never the premium budget.
                if status == "trialing" and not _TRIAL_GRANTS_FULL_PLAN and plan == "pro":
                    return "standard"
                return plan
    if invited:
        return "standard"
    return None


def resolve_stage2_model(plan, requested_model, ent=None):
    """The entitled stage-2 model: the requested one if the plan grants its slot,
    else CHEAP_MODEL. `ent` overrides the compiled ENTITLEMENTS map (T1 overlay)."""
    ent = ent if ent is not None else ENTITLEMENTS
    if plan:
        slot = model_slot(requested_model)
        if slot and ent[plan]["stage2_models"].get(slot) is not None:
            return requested_model
    return CHEAP_MODEL


def daily_review_cap(plan, model, ent=None):
    """Per-user, per-day review cap for (plan, resolved stage-2 model). None -> 0.
    `ent` overrides the compiled ENTITLEMENTS map (T1 overlay)."""
    if not plan:
        return 0
    ent = ent if ent is not None else ENTITLEMENTS
    caps = ent[plan]["stage2_models"]
    slot = model_slot(model) or "cheap"
    cap = caps.get(slot)
    if cap is None:
        cap = caps.get("cheap", 0)
    return cap


def monthly_allowance(plan, kind, ent=None):
    """Monthly generation allowance for kind in ('resume','cover'). None -> 0.
    `ent` overrides the compiled ENTITLEMENTS map (T1 overlay)."""
    if not plan:
        return 0
    ent = ent if ent is not None else ENTITLEMENTS
    e = ent[plan]
    return e["monthly_resume"] if kind == "resume" else e["monthly_cover"]
